core/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java

```diff
@@ -0,0 +1,263 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.shard;
+
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.index.IndexWriterConfig;
+import org.apache.lucene.index.SegmentInfos;
+import org.apache.lucene.store.Directory;
+import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.routing.RestoreSource;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.lucene.Lucene;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.index.engine.EngineException;
+import org.elasticsearch.index.snapshots.IndexShardRepository;
+import org.elasticsearch.index.snapshots.IndexShardRestoreFailedException;
+import org.elasticsearch.index.store.Store;
+import org.elasticsearch.indices.recovery.RecoveryState;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.common.unit.TimeValue.timeValueMillis;
+
+/**
+ * This package private utility class encapsulates the logic to recover an index shard from either and existing index on
+ * disk or from a snapshot in a repository.
+ */
+final class StoreRecovery {
+
+    private final ESLogger logger;
+    private final ShardId shardId;
+
+    StoreRecovery(ShardId shardId, ESLogger logger) {
+        this.logger = logger;
+        this.shardId = shardId;
+    }
+
+    boolean recoverFromStore(final IndexShard indexShard, final boolean indexShouldExists, DiscoveryNode localNode) {
+        if (canRecover(indexShard)) {
+            if (indexShard.routingEntry().restoreSource() != null) {
+                throw new IllegalStateException("can't recover - restore source is not null");
+            }
+            try {
+                indexShard.recovering("from store", RecoveryState.Type.STORE, localNode);
+            } catch (IllegalIndexShardStateException e) {
+                // that's fine, since we might be called concurrently, just ignore this, we are already recovering
+                return false;
+            }
+            return recover(indexShard, () -> {
+                logger.debug("starting recovery from shard_store ...");
+                RecoveryState recoveryState = indexShard.recoveryState();
+                runRecoverFromStore(indexShard, indexShouldExists, recoveryState);
+                return recoveryState;
+            });
+        }
+        return false;
+    }
+
+    boolean restoreFromRepository(final IndexShard indexShard, IndexShardRepository repository) {
```

Comment:
I think the namespace just confuses the difference between "restore" and "recover", maybe some javadoc for the class to clarify that?